crates/resolver/src/name_server/name_server_stats.rs

```diff
@@ -6,125 +6,295 @@
+   /// follows:
+   ///
+   /// 1. The value is initialized to a random value that represents a very low
+   ///    latency.
+   /// 2. If the round-trip time (RTT) was successfully measured for a query,
+   ///    then it is incorporated into the EWMA using the formula linked above.
+   /// 3. If the RTT could not be measured (i.e. due to a connection failure),
+   ///    then a constant penalty factor is applied to the EWMA.
+   /// 4. When comparing EWMA values, a time-based decay is applied to each
+   ///    value. Note that this decay is only applied at read time.
+   ///
+   /// For the original discussion regarding this algorithm, see
+   /// https://github.com/bluejekyll/trust-dns/issues/1702.
+   srtt_microseconds: AtomicU32,
+
+
+   /// The last time the `srtt_microseconds` value was updated.
+   last_update: Arc<Mutex<Option<Instant>>>,
 }
 
 impl Default for NameServerStats {
-    fn default() -> Self {
-        Self::new(0, 0)
-    }
+   fn default() -> Self {
+       // Initialize the SRTT to a randomly generated value that represents a
+       // very low RTT. Such a value helps ensure that each server is attempted
+       // early.
+       Self::new(rand::thread_rng().gen_range(1..32))
+   }
+}
+
+/// Returns an exponentially weighted value in the range of 0.0 < x < 1.0
+///
+/// Computes the value using the following formula:
+///
+/// e<sup>(-t<sub>now</sub> - t<sub>last</sub>) / weight</sup>
+///
+/// As the duration since the `last_update` approaches the provided `weight`,
+/// the returned value decreases.
+fn compute_srtt_factor(last_update: Instant, weight: u32) -> f64 {
+   let exponent = (-last_update.elapsed().as_secs_f64().max(1.0)) / f64::from(weight);
+   exponent.exp()
 }
 
 impl NameServerStats {
-    pub(crate) fn new(successes: usize, failures: usize) -> Self {
-        Self {
-            successes: AtomicUsize::new(successes),
-            failures: AtomicUsize::new(failures),
-        }
-    }
-
-    pub(crate) fn next_success(&self) {
-        self.successes.fetch_add(1, atomic::Ordering::Release);
-    }
-
-    pub(crate) fn next_failure(&self) {
-        self.failures.fetch_add(1, atomic::Ordering::Release);
-    }
-
-    fn noload_eq(
-        self_successes: usize,
-        other_successes: usize,
-        self_failures: usize,
-        other_failures: usize,
-    ) -> bool {
-        self_successes == other_successes && self_failures == other_failures
-    }
+   const CONNECTION_FAILURE_PENALTY: u32 = Duration::from_millis(150).as_micros() as u32;
+   const MAX_SRTT_MICROS: u32 = Duration::from_secs(5).as_micros() as u32;
+
+   pub(crate) fn new(initial_srtt: u32) -> Self {
+       Self {
+           srtt_microseconds: AtomicU32::new(initial_srtt),
+           last_update: Arc::new(Mutex::new(None)),
+       }
+   }
+
+   /// Records the measured `rtt` for a particular query.
+   pub(crate) fn record_rtt(&self, rtt: Duration) {
```

Comment:
Sorry, not sure I follow. How would this work? The scope of the returned value seems like it would just be `NameServer::inner_send`, which would mean that we'd always record a connection failure?

Reply:
Yeah, it probably doesn't make sense with the structure here. I thought it might be nice to guard against a scenario where we fail to record any SRTT for a given request.

Reply:
Thanks for explaining! Marking this as resolved since no further action is needed.